### PR TITLE
Bluetooth: Mesh: Mesh BLOB Transfer sample

### PR DESCRIPTION
**Pull request by tot0rokr**
Sample code that uses BLOB Transfer Server/Client models.
This sample has been tested with the nRF52840 DK (nrf52840dk_nrf52840) board.

DevZone Case ID: 331920, 334530

### CODE REVIEW

File: samples/bluetooth/mesh/blob_transfer/src/blob_shell.c [hunk 1]
```diff
@@ -0,0 +1,542 @@
...
+{
+	return _cmd_blob(read_flash_area, shell, argc, argv);
+}
+
+inline static int cmd_write_blob(const struct shell *shell, size_t argc, char *argv[])
+{
+	return _cmd_blob(write_flash_area, shell, argc, argv);
+}
+
+inline static int cmd_hash_blob(const struct shell *shell, size_t argc, char *argv[])
```
Comment (omkar3141) [inline samples/bluetooth/mesh/blob_transfer/src/blob_shell.c:509]: We sort of have all of this already:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/shell/blob.c#L569 

Couldn't you use that?

File: samples/bluetooth/mesh/blob_transfer/README.rst [hunk 2]
```diff
@@ -0,0 +1,310 @@
+.. _ble_mesh_blob_xfer:
+
+Bluetooth Mesh: Binary Large Object (BLOB) Transfer
```
Comment (omkar3141) [inline samples/bluetooth/mesh/blob_transfer/README.rst:3]: If you use existing `mesh_shell` test application, and enable BLOB srv and cli models there. Then you can evaluate the BLOB transfer like this:

On BLOB sender shell:
```
mesh init
mesh cdb create 1234
mesh cdb show
mesh cdb app-key-add 0 0 
mesh prov local 0 1
mesh models cfg appkey add 0 0
mesh models cfg model app-bind 1 0 0x1400
mesh models cfg model app-bind 1 0 0x1401
mesh prov auth-method none
mesh prov remote-adv dddd0000000000000000000000000000 0 5 10
mesh target dst 5
mesh models cfg appkey add 0 0
mesh models cfg model app-bind 5 0 0x1401
mesh models cfg model app-bind 5 0 0x1400
mesh models blob cli target 0x05
```

On BLOB receiver shell:
```
mesh models blob srv rx 0xac
```

Then, again BLOB sender shell:
```
mesh models blob cli tx 0xac 1024 12 30
```

File: samples/bluetooth/mesh/blob_transfer/src/blob_shell.c (hunk 1 above)
Comment (tot0rokr) [reply samples/bluetooth/mesh/blob_transfer/src/blob_shell.c:509]: I misunderstood the conversation we had on DevZone. I'll fix it by using the implementation you mentioned.

File: samples/bluetooth/mesh/blob_transfer/README.rst (hunk 2 above)
Comment (tot0rokr) [reply samples/bluetooth/mesh/blob_transfer/README.rst:3]: Okay. I'll merge and configure only the parts that are not in the `mesh_shell` test.

File: samples/bluetooth/mesh/blob_transfer/README.rst [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+
+* A :file:`src/main.c` file to handle Bluetooth Mesh initialization.
+* File :file:`src/model_handler.c` including the model handling for Device Composition Data, Health and Configuration Server models.
+* File :file:`src/blob_xfer.c` with the BLOB Transfer Server/Client model handlers and APIs of Sender/Receiver implementation.
+* File :file:`src/blob_shell.c` with the :ref:`shell module <shell_api>` for communicating with users.
+
+Partition Manager
+=================
+
+This sample use static configuration for Partition Manager.
```
Comment (nordicjm) [inline samples/bluetooth/mesh/blob_transfer/README.rst:97]:
```suggestion
This sample uses a static Partition Manager configuration.
```

File: samples/bluetooth/mesh/blob_transfer/src/blob_xfer.c [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+	hash_flash_area(blob_tx.blob_tx_flash_stream.area_id, blob_tx.blob_tx_flash_stream.offset,
+			blob_tx.xfer.size);
+
+	return bt_mesh_blob_cli_caps_get(blob_cli, &blob_tx.inputs);
+}
+
+int blob_tx_cancel(void)
+{
+	if (!bt_mesh_blob_cli_is_busy(blob_cli))
+		return -1;
```
Comment (nordicjm) [inline samples/bluetooth/mesh/blob_transfer/src/blob_xfer.c:368]: use brackets

File: samples/bluetooth/mesh/blob_transfer/src/blob_xfer.c [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+
+int blob_rx_flash_init(off_t offset)
+{
+	int err = 0;
+
+	err = bt_mesh_blob_io_flash_init(&blob_rx_flash_stream, BLOB_AREA_ID,
+					 ROUND_DOWN(offset, 4));
+
+	if (err)
+		return err;
```
Comment (nordicjm) [inline samples/bluetooth/mesh/blob_transfer/src/blob_xfer.c:393]: and for all cases

File: samples/bluetooth/mesh/blob_transfer/include/blob_xfer.h [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define MAX_TARGET_NODES CONFIG_BT_MESH_BLOB_TX_MAX_TARGETS
+
+#define STORAGE_START 0x0000
+#define STORAGE_END 0x8000
+#define STORAGE_ALIGN 0x1000
```
Comment (nordicjm) [inline samples/bluetooth/mesh/blob_transfer/include/blob_xfer.h:25]: HUH??? Use PM defines for these

**Review by nordicjm — CHANGES_REQUESTED** [review]

**Comment by HaavardRei** [comment]
Hi @tot0rokr,
On behalf of the ncs Mesh team, we appreciate the time and effort you have put into this PR, but we will not merge this sample into sdk-nrf. The reason for this is that the functionality of the the BLOB models is demonstrated and tested using the Mesh DFU samples, and as @omkar3141 showed it is possible to evaluate only the BLOB models using the [mesh_shell test application](https://github.com/nrfconnect/sdk-zephyr/tree/main/tests/bluetooth/mesh_shell). 

I would suggest that you try it out using this. If you feel that any functionality is lacking or need some assistance feel free to reach out (here or creating a zephyr issue), or even making an upstream PR augmenting the shell :)

**Comment by tot0rokr** [comment]
Hi @HaavardRei,

It seems quite challenging to derive a clear usage example of the BLOB Transfer model from the DFU sample.

As mentioned by @omkar3141, most of what I implemented has already been evaluated within the Mesh Shell Test app. Originally, I intended to implement something simpler, but as I continued to refine it, I ended up creating a much larger example than planned.

However, I created this sample because I found it much harder to understand how to use the BLOB Trasnfer model with the Mesh Shell Test app and DFU examples than with other, more simple samples, and I’ve seen that there is demand for such guidance on DevZone.

If merging this sample isn’t possible, I’d welcome other ideas. For example, it would be very helpful to have a dedicated sample that demonstrates both simple flash read/write operations and basic PM operations, and would avoid analyzing unnecessary DFU details.
I’d like to propose creating a more streamlined example that focuses solely on flash operations without shell functionalities.

**Comment by HaavardRei** [comment]
I agree that the documentation for using only the BLOB transfer models is not optimal. I've already created some upstream zephyr PRs for updating documentation for the shell commands, and I've also added a task to our backlog for adding a more detailed guide on using the BLOB models and reading/writing to flash.

**Comment by tot0rokr** [comment]
@HaavardRei Okay. Then, I think this PR can be closed.

**Pull request closed without merging**